heron/packing/src/java/org/apache/heron/packing/binpacking/TwoDimensionVectorBinPacking.java

```diff
@@ -0,0 +1,379 @@
+ * <p>
+ * 6. The disk required for one instance is calculated as the default disk value for one instance.
+ * <p>
+ * 7. The RAM required for a container is calculated as:
+ * (RAM for instances in container) + (paddingPercentage * RAM for instances in container)
+ * <p>
+ * 8. The CPU required for a container is calculated as:
+ * (CPU for instances in container) + (paddingPercentage * CPU for instances in container)
+ * <p>
+ * 9. The disk required for a container is calculated as:
+ * (disk for instances in container) + ((paddingPercentage * disk for instances in container)
+ * <p>
+ * 10. The pack() return null if PackingPlan fails to pass the safe check, for instance,
+ * the size of RAM for an instance is less than the minimal required value.
+ */
+public class TwoDimensionVectorBinPacking extends AbstractPacking {
+
+  private static final Logger LOG = Logger.getLogger(TwoDimensionVectorBinPacking.class.getName());
+
+  private int numContainers = 0;
+
+  private PackingPlanBuilder newPackingPlanBuilder(PackingPlan existingPackingPlan) {
+    return new PackingPlanBuilder(topology.getId(), existingPackingPlan)
+        .setDefaultInstanceResource(defaultInstanceResources)
+        .setMaxContainerResource(maxContainerResources)
+        .setRequestedContainerPadding(padding)
+        .setRequestedComponentResource(componentResourceMap)
+        .setInstanceConstraints(Arrays.asList(new MinRamConstraint(), new MinCpuConstraint()))
+        .setPackingConstraints(Arrays.asList(new ResourceConstraint(),
+            new InstanceDensityConstraint(maxNumInstancesPerContainer)));
+  }
+
+  /**
+   * Get a packing plan using First Fit Decreasing
+   *
+   * @return packing plan
+   */
+  @Override
+  public PackingPlan pack() {
+    PackingPlanBuilder ramFirstPlanBuilder = newPackingPlanBuilder(null);
+    PackingPlanBuilder cpuFirstPlanBuilder = newPackingPlanBuilder(null);
+
+    // Get the instances using FFD allocation
+    try {
+      numContainers = 0;
+      ramFirstPlanBuilder = get2DVectorBinAllocation(ramFirstPlanBuilder,
+          SortingStrategy.RAM_FIRST);
+      numContainers = 0;
+      cpuFirstPlanBuilder = get2DVectorBinAllocation(cpuFirstPlanBuilder,
+          SortingStrategy.CPU_FIRST);
+
+      PackingPlan ramFirstPlan = ramFirstPlanBuilder.build();
+      PackingPlan cpuFirstPlan = cpuFirstPlanBuilder.build();
+
+      return pickBetter(ramFirstPlan, cpuFirstPlan);
+    } catch (ConstraintViolationException e) {
+      throw new PackingException("Could not allocate all instances to packing plan", e);
+    }
+  }
+
+  /**
+   * Pick the better plan by checking which plan dominates in terms of:
+   * overall resource utility, average container resource utility,
+   * container ram utility stdev, and container cpu utility stdev.
+   * @param plan1
+   * @param plan2
+   * @return the better plan
+   */
+  private PackingPlan pickBetter(PackingPlan plan1, PackingPlan plan2) {
```

Comment:
Could we make this function a static function? I am feeling that we might move this function to a util or some other class in future.